### ISSUE TITLE
special characters in channel names cause error

### ISSUE THREAD

**Issue by jchris** [couchbase/sync_gateway#656]
this is a bug when you are trying to generate channel names from user data.

**Comment by deefactorial** [comment]
https://groups.google.com/forum/#!topic/mobile-couchbase/C2GjqSWy9oA

**Comment by deefactorial** [comment]
My work around is to base 64 encode the user data for the channel name.

**Comment by zgramana** [comment]
We need to document this once fixed.

**Comment by zgramana** [comment]
As part of this, we need to canonicalize the set of valid chars for document ids and channel names, since the former are often consumed in the later. The charset also needs to be URL safe, as the REST GET/PUT/DELETE for a specific document includes the doc id as part of the URL.

**Comment by deefactorial** [comment]
Is it possible to URL encode special characters for document ids?

**Comment by snej** [comment]
It was also pointed out by a forum user that email addresses can't safely be used as channel names because they [allow a lot of characters](https://en.wikipedia.org/wiki/Email_address#Syntax) that aren't allowed in channels.

At a minimum, if we added `'%'` to the channel character set then arbitrary strings could be hex-encoded as per URLs. (The Wikipedia page I linked above says that this encoding can also be used in email addresses.)

This limited character set is my fault, but honestly I can't remember why i was so restrictive about it. The only problem I can think of with an arbitrary character set would be losing the ability to concatenate channel names in a plain string, since there wouldn't be any delimiter you could put between them; but I don't think we ever do that since a JSON array is such an obvious way to express that.

**Comment by deefactorial** [comment]
My work around was to base64 encode the channel names, although difficult
to read and not that efficient it works.

**Comment by jchris** [comment]
IMHO the allowed channel names should be all UTF-8 strings, otherwise sync functions fill with BS.

**Comment by jchris** [comment]
Or is it UTF-16? Whatever it is we should strive for unsurprising.

**Comment by jchris** [comment]
I am running into this again. I really do think the allowed set of channel names should be all unicode strings. The [PR looks trivial.](https://github.com/couchbase/sync_gateway/blob/master/src/github.com/couchbase/sync_gateway/channels/set.go#L36)

**Comment by jchris** [comment]
For the immediate term, we can make the only character not allowed in channel names be `,`.  In the longer term we can change that but it will require coordination with Couchbase Lite, because the channels filter param is passed as a [comma delimited string](https://github.com/couchbase/sync_gateway/blob/master/src/github.com/couchbase/sync_gateway/rest/changes_api.go#L94) not a JSON array, in the rest API currently. There are no technical hurdles to changing that, except we need to be able to support the current client behavior for the rest of the release lifetime.

**Comment by jchris** [comment]
Questions:
- what about the empty string? I'm coding it as invalid.
- just double checking that we are ok with having these be valid: `"*", "**", "a", "a ", "a b", "a*b", "FOO", "123", "-z", "foo_bar", "Éclær", "z7_", "!",  "Z∫•", "*!"`

**Comment by adamcfraser** [comment]
`"*" and "!"` should be prohibited (used for the admin and public channels), as well as empty string.

There's another ticket that proposes using the `!` to identify named public channels (https://github.com/couchbase/sync_gateway/issues/947), so at minimum we should restrict that as an opening channel name.

Does this change expose us to any type of injection attack when the channel name gets used as part of a view query (e.g. when it's included as part of the startkey in a channel backfill request)?

**Comment by jchris** [comment]
View query keys are any string, so that should be fine. I think reserving the ! and \* names is fine, but [they are valid today.](https://github.com/couchbase/sync_gateway/blob/master/src/github.com/couchbase/sync_gateway/channels/set_test.go#L19)

I think `!example` should be valid, also `*example` and `_foo`

**Comment by jchris** [comment]
Basically, channel names should be subject to the same restrictions as view keys, (none) but for the short term we have to keep the comma restriction in there until we deprecate the current clients query format

**Comment by adamcfraser** [comment]
I guess I should clarify.  `!` and `*` are valid channel names (and should continue to be), but they have custom functionality associated with them.  The same would be true for `!example` after #947 gets implemented.

So I guess my concern isn't whether these should be allowed, but the underlying implication of creating channels based on user data.  App developers who don't want users to be able to make documents public will need to ensure they aren't allowing docs to be assigned to channel `!` (or `!example` in future).  That's not any different than where we are today, but might merit some documentation/recommendations around usage.

**Comment by jchris** [comment]
I think we decided that we wouldn't want to do channel namespaces using special character prefixes, because we want to allow user data as channel names.

I got bit by this again trying to use Stack Overflow api data tags as channel names, but some questions are tagged `c#`.

**Comment by tleyden** [comment]
What are the currently allowed characters in channel names?  (pre https://github.com/couchbase/sync_gateway/pull/1293, since it hasn't been merged)

Specifically are `:` and `-` allowed?

**Comment by pasin** [comment]
Based on my test, '-' is allowed but not ':'.

**Comment by jamescooke** [comment]
Agree with @pasin . However the docs for 1.2 don't show this http://developer.couchbase.com/documentation/mobile/1.2/develop/guides/sync-gateway/channels/index.html
